src/librustc/middle/resolve_lifetime.rs

```diff
@@ -2154,7 +2154,34 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
                 false
             };
 
-            if let hir::TyKind::Rptr(lifetime_ref, ref mt) = inputs[0].node {
+            let mut self_arg = &inputs[0].node;
+
+            // Apply `self: &(mut) Self` elision rules even if nested in `Pin`.
+            loop {
+                if let hir::TyKind::Path(hir::QPath::Resolved(None, ref path)) = *self_arg {
+                    if let Res::Def(DefKind::Struct, def_id) = path.res {
+                        if self.tcx.lang_items().pin_type() == Some(def_id) {
```

Comment:
Wait... are we doing this specifically for `Pin`? Why?

Reply:
@Centril

Based on https://github.com/rust-lang/rust/pull/60944#issuecomment-494722365 (but I may have missed something else...).

> Could we special-case `Pin<&mut Self>`? Or more generally, use `self: &mut Self` elision rules even if nested in a `Pin`?
> The fully general case could be a nightmare to untangle, likely for any particular behavior set we may pick.

Also, [other stable receiver types cannot take `&(mut) Self` without unstable feature.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=30b49846e6f0d54a1f1329dff2b353d5)

Reply:
I certainly don't think this should query for `.pin_type()`; that seems like a hack. If we do anything like this, then it should be formulated generally for places where `&mut? S` where `is_self_ty(S)` can occur possibly with some logic for how to handle cases with multiple `&mut? S` occurrences in the `self` type (e.g. 1) give up and error, 2) pick the first mention).

Reply:
Yeah, I also think this is a hack. Fixed.
(By the way, [the multiple `&mut? S` you said are not supported without unstable feature](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a899ccb1fbe5f476fcbce2f84e0a22c6))

Reply:
>  (By the way, [the multiple `&mut? S` you said are not supported without unstable feature](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a899ccb1fbe5f476fcbce2f84e0a22c6))

That doesn't seem important -- knowing that there is a general rule is. We should also have tests for the new logic + `arbitrary_self_types`, but you may want to wait with investing more time in this until we figure out the general direction.